feat: add attachments to notification

HenningWendtland opened the pull request:
**Business Case:**
My client wants to send out Sales Invoices automatically at some point after submission, using **Notifications**.  
The print is generated with [PDF on Submit](https://github.com/alyf-de/erpnext_pdf-on-submit) and depends on different settings (print formats, letterheads) and also should include an [e-Invoice](https://github.com/alyf-de/eu_einvoice).
The print generated by the _Attach Print_ checkbox in **Notification** is therefore not sufficient and cannot be used. Instead, we want to send files that are either general attachments to the doc, or attached to a specific field.

**Implementation:**
- Add a new "Select" field called _Attach Files_ to **Notification** with options "" (empty), "From Field" and "All".
	- If "From Field" is selected, we show an additional "Select" field called _From Attach Field_, where the user can select the specific "Attach" type field from the reference DocType from which to take the file.
	- If "All" is selected, we send all files attached to the doc.
- The previous _Attach Print_ functionality is slightly refactored but remains essentially unchanged.
- The two settings for attaching print and files can be combined.

---

Attach print, works as before:

<img width="935" height="185" alt="print" src="https://github.com/user-attachments/assets/7b89f28e-c217-4cab-84cf-b2d24782bce8" />


---

Add all attachments from doc to email:

<img width="935" height="146" alt="all_files" src="https://github.com/user-attachments/assets/b5e76308-d1de-4d0d-a45a-96162eb56506" />

---

Add attachments from specific docfield to email:

<img width="935" height="219" alt="from_field" src="https://github.com/user-attachments/assets/f8c6a884-685c-47e9-bc98-ea1b295e76bd" />


Docs: https://docs.frappe.io/framework/notifications#attachment-settings

Reference: IBT-91

HenningWendtland committed:
feat: add attachments to notification

HenningWendtland committed:
chore: remove unneeded comment

barredterra commented on frappe/email/doctype/notification/notification.py:
```diff
@@ -624,10 +626,25 @@ def get_receiver_list(self, doc, context, field_on_user="mobile_no", recipient_e
-		"""check print settings are attach the pdf"""
-		if not self.attach_print:
-			return None
+	def get_attachments(self, doc) -> list[dict]:
```
```suggestion
	def get_attachment(self, doc) -> list[dict]:
```

While the proposed name is clearly better, let's not change it in order to maintain backwards compatibility.

barredterra commented on frappe/email/doctype/notification/notification.py:
```diff
@@ -624,10 +626,25 @@ def get_receiver_list(self, doc, context, field_on_user="mobile_no", recipient_e
+	def get_attachments(self, doc) -> list[dict]:
+		"""Check Attachment Settings and return attachments accordingly"""
+		attachments = []
+		if self.attach_print and (print_info := self.get_print(doc)):
+			attachments.append(print_info)
```
```suggestion
		if self.attach_print:
			attachments.append(self.get_print(doc))
```

This is not really needed as part of the if-statement, because `get_print` will never return a falsy value.

barredterra commented on frappe/email/doctype/notification/notification.py:
```diff
@@ -642,18 +659,16 @@ def get_attachment(self, doc):
+				"name": doc.name,
+				"print_format": self.print_format,
+				"print_letterhead": print_settings.with_letterhead,
+				"lang": frappe.db.get_value("Print Format", self.print_format, "default_print_language")
```
```suggestion
				"lang": doc.get("language") or frappe.db.get_value("Print Format", self.print_format, "default_print_language")
```

Not related to this PR, but using the specific transaction's language with highest priority would be correct. Maybe we can sneak this in here 😁 (or create a separate PR later)

HenningWendtland committed:
fix: revert change of function name

HenningWendtland committed:
fix: simplify if statement

HenningWendtland committed:
fix: use doc language with highest prio in print

barredterra committed:
Merge remote-tracking branch 'upstream/develop' into add-attachments-to-notification

barredterra commented:
Tested:

- [x] Attachments (print and from field) are sent as expected in an email notification

Todo:

- [x] Check letterhead param in sendmail
- [x] Add test cases

barredterra committed:
fix: prevent empty list in attached_file

For empty list []:
  - attachments and json.dumps(attachments) → returns [] (the empty list, because and short-circuits on falsy left side)
  - json.dumps(attachments) if attachments else None → returns None (because empty list is falsy in if condition)

barredterra committed:
fix(Notification): improve print letterhead handling in email attachments

Fix the logic for determining the print letterhead attachment to handle cases with multiple attachments, and only one of them printed.

barredterra committed:
refactor: get_attachment to improve clarity

barredterra commented:
LGMT now 👍🏼

barredterra committed:
fix(Notification): better fieldnames

barredterra committed:
fix(Notification): condition & validation for From Attach Field

barredterra committed:
test(Notification): attach files

barredterra committed:
Merge branch 'develop' into add-attachments-to-notification

barredterra commented:
The linter failure is due to usage of `frappe.db.commit()` in a context manager for the test cases. This pattern was adopted from an existing context manager in the same file.

iamejaaz commented:
@HenningWendtland  Time to update docs 👀

barredterra commented:
@iamejaaz the docs site is still missing login and edit buttons. Can't really expect people to add docs if it's that difficult. Also, my impression is that nobody is reviewing suggestions there.

https://docs.frappe.io/framework/notifications#attachment-settings
